src/ray/core_worker/core_worker.cc

```diff
@@ -944,7 +950,11 @@ Status CoreWorker::Create(const std::shared_ptr<Buffer> &metadata, const size_t
 
 Status CoreWorker::Seal(const ObjectID &object_id, bool pin_object,
                         const absl::optional<rpc::Address> &owner_address) {
-  RAY_RETURN_NOT_OK(plasma_store_provider_->Seal(object_id));
+  auto status = plasma_store_provider_->Seal(object_id);
+  if (!status.ok()) {
+    reference_counter_->RemoveOwnedObject(object_id);
```

Comment:
Is this guaranteed to only get called by the owner of the object?

Reply:
It was my impression that only owned objects are `Seal`ed, some time after a `Create()`; is that not the case?

Reply:
It looks like it also gets called for object creation in general, like when a worker stores task outputs: https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L1837

Reply:
Yes, but I think that's under the expectation that the language frontend [will make the `Create()` call](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/python/ray/_raylet.pyx#L1482-L1485) via [`CoreWorker::AllocateReturnObjects`](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/src/ray/core_worker/core_worker.cc#L1752-L1753), at least for Python.

Reply:
I took [this bit](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/src/ray/core_worker/core_worker.h#L525-L526) of the `Seal()` method documentation to imply that `Create()` proceeding `Seal()` is a necessary precondition that callers are expected to adhere to.

Reply:
What I was missing here is that the `Create()` call is the [_other_ `CoreWorker::Create()` method](https://github.com/ray-project/ray/blob/4aef3d68360226bfaf3fa2fa8e425f4ef91873cb/src/ray/core_worker/core_worker.cc#L933-L943), which doesn't add an owned reference. Meanwhile, the same `Seal()` method is used by both.